### ISSUE TITLE
Damage/rendering issue with window borders

### ISSUE THREAD

**Issue by BitterlyIronic** [labwc/labwc#501]
I've been seeing an issue with window borders not rendering correctly when overlapping other windows (or their title bars, as seen below).

The theme being used is [Nightmare](https://www.box-look.org/p/1017737), and the issue is entirely localized to the window borders (e.g., there are no damage issues when the border width is set to 0).

With WLR_SCENE_DEBUG_DAMAGE=highlight:
![image](https://user-images.githubusercontent.com/44883930/186513522-ed13d871-9bdf-4445-8b5b-45419a4e08aa.png)
Without (it actually only does this when slurp is running, normally the border just goes transparent): 
![image](https://user-images.githubusercontent.com/44883930/186513765-fde8a524-3c20-4810-85b5-942a619ebedd.png)

**Comment by johanmalm** [comment]
Thanks for reporting. Is the issue specific to that theme?

**Comment by BitterlyIronic** [comment]
It does seem like it's something specific something Nightmare is doing. [Greybird](https://www.box-look.org/p/1017695), as a random example, looks like it's rendering correctly (this is with WLR_SCENE_DEBUG_DAMAGE):

![image](https://user-images.githubusercontent.com/44883930/186517825-88832279-099c-407e-8eda-b934c1f07ce3.png)

**Comment by BitterlyIronic** [comment]
Oh, additional data point I forgot to mention: This does not happen on the 0.5.3 release, only when running git HEAD.

**Comment by johanmalm** [comment]
Weird. I can't reproduce it. It is only the top border?

**Comment by BitterlyIronic** [comment]
It happens on every border (except for around the buttons) and on further investigation it happens specifically for a border color of #000000 for whatever reason. I changed it to #010101 and it stopped happening, and I set it back to #000000, and it started right back up.

With WLR_SCENE_DEBUG_DAMAGE and borders set to #010101
![image](https://user-images.githubusercontent.com/44883930/186532824-0bee0197-df87-4fc6-8f45-7e55ea70b421.png)

In conclusion, I have no idea.

**Comment by Consolatis** [comment]
There is [a commit](https://gitlab.freedesktop.org/wlroots/wlroots/-/commit/0a9a941ba329d10c08a6f638a0b22b60b7c92d0f) (+later ones) in wlroots latest scene optimizations which detects black areas and optimizes something based on that (didn't read it in depth). That feature is still worked on (see [the latest commits](https://gitlab.freedesktop.org/wlroots/wlroots/-/commits/master/types/scene/wlr_scene.c)).

@johanmalm maybe it can be reproduced with updating subprojects/wlroots to latest master. It might also have been fixed in wlroots in the mean time.
@BitterlyIronic does setting `WLR_SCENE_DISABLE_VISIBILITY=1` or updating wlroots fix it?

**Comment by BitterlyIronic** [comment]
Setting WLR_SCENE_DISABLE_VISIBILITY does fix it! (I'm already running the latest commit of wlroots).

**Comment by Consolatis** [comment]
It will likely happen as well when setting `window.active.title.bg.color` to `#000000` as well. Seems like a bug in wlroots.

**Comment by Nefsen402** [comment]
@BitterlyIronic 

With the black rect, does everything where the black rect appear to be junk that just changes all the time and is completely random or is it showing through to a surface underneath?

**Comment by BitterlyIronic** [comment]
@Nefsen402

It's the color of the window underneath, I first noticed it because the edges of my wallpaper were showing through when I had a window maximized.

**Comment by Nefsen402** [comment]
@BitterlyIronic Please confirm this diff for wlroots fixes your issue.

```
diff --git a/types/scene/wlr_scene.c b/types/scene/wlr_scene.c
index 102e9e3ab..2682ff2ae 100644
--- a/types/scene/wlr_scene.c
+++ b/types/scene/wlr_scene.c
@@ -221,8 +221,8 @@ static bool scene_nodes_in_box(struct wlr_scene_node *node, struct wlr_box *box,
        return _scene_nodes_in_box(node, box, iterator, user_data, x, y);
 }
 
-static void scene_node_cull_hidden(struct wlr_scene_node *node, int x, int y,
-               pixman_region32_t *visible) {
+static void scene_node_opaque_region(struct wlr_scene_node *node, int x, int y,
+               pixman_region32_t *opaque) {
        if (node->type == WLR_SCENE_NODE_RECT) {
                struct wlr_scene_rect *scene_rect = scene_rect_from_node(node);
                if (scene_rect->color[3] != 1) {
@@ -236,19 +236,16 @@ static void scene_node_cull_hidden(struct wlr_scene_node *node, int x, int y,
                }
 
                if (!buffer_is_opaque(scene_buffer->buffer)) {
-                       pixman_region32_translate(visible, -x, -y);
-                       pixman_region32_subtract(visible, visible, &scene_buffer->opaque_region);
-                       pixman_region32_translate(visible, x, y);
+                       pixman_region32_copy(opaque, &scene_buffer->opaque_region);
+                       pixman_region32_translate(opaque, x, y);
                        return;
                }
        }
 
        int width, height;
        scene_node_get_size(node, &width, &height);
-       pixman_region32_t opaque;
-       pixman_region32_init_rect(&opaque, x, y, width, height);
-       pixman_region32_subtract(visible, visible, &opaque);
-       pixman_region32_fini(&opaque);
+       pixman_region32_fini(opaque);
+       pixman_region32_init_rect(opaque, x, y, width, height);
 }
 
 struct scene_update_data {
@@ -370,7 +367,11 @@ static bool scene_node_update_iterator(struct wlr_scene_node *node,
                lx, ly, box.width, box.height);
 
        if (data->calculate_visibility) {
-               scene_node_cull_hidden(node, lx, ly, data->visible);
+               pixman_region32_t opaque;
+               pixman_region32_init(&opaque);
+               scene_node_opaque_region(node, lx, ly, &opaque);
+               pixman_region32_subtract(data->visible, data->visible, &opaque);
+               pixman_region32_fini(&opaque);
        }
 
        update_node_update_outputs(node, data->outputs, NULL);
@@ -1508,7 +1509,12 @@ bool wlr_scene_output_commit(struct wlr_scene_output *scene_output) {
                        struct wlr_scene_node *node = list_data[i];
                        int x, y;
                        wlr_scene_node_coords(node, &x, &y);
-                       scene_node_cull_hidden(node, x, y, &background);
+                       pixman_region32_t opaque;
+                       pixman_region32_init(&opaque);
+                       scene_node_opaque_region(node, x, y, &opaque);
+                       pixman_region32_intersect(&opaque, &opaque, &node->visible);
+                       pixman_region32_subtract(&background, &background, &opaque);
+                       pixman_region32_fini(&opaque);
                }
        }
 ```

**Comment by Nefsen402** [comment]
I also created a merge request here: https://gitlab.freedesktop.org/wlroots/wlroots/-/merge_requests/3679

**Comment by BitterlyIronic** [comment]
> @BitterlyIronic Please confirm this diff for wlroots fixes your issue.
> 
> [fix.txt](https://github.com/labwc/labwc/files/9420201/fix.txt)

Yep, that does it. WLR_SCENE_DEBUG_DAMAGE set and pure black borders and everything looks good with that patch applied.

![image](https://user-images.githubusercontent.com/44883930/186546952-2a7a23a1-9824-428a-9c7a-b281a3c48ea6.png)

Thanks for the quick turnaround!

**Comment by Consolatis** [comment]
@Nefsen402 thanks a lot for your help.

**Comment by kode54** [comment]
Would this issue also be causing window titles to never update? I'm having an issue where window titles stretch and unstretch depending on the text length, but the surface of the text remains what it was when the window was created, and just stretches to fill the space.

**Comment by Nefsen402** [comment]
@kode54 That sounds like an entirely different issue, please consider opening another issue and documenting it.
